joomla/joomla-cms
# [4.x] More flexible display of relative dates.

### Summary of Changes

This is an alternative solution to #31675 and fixes the concerns discussed in https://github.com/joomla/joomla-cms/pull/32911#issuecomment-809301133 and the following comments.
This PR introduces a new HTMLHelper method called `date.relativeFormatted` which can be used to display dates in a relative format without losing important information. 

Key features:
- Relative or absolute format is chosen based on user preference.
- Using the flag `$forceRelative`, relative display can be forced, ignoring the user's preference (use carefully!).
- In addition to the relative date display, the full absolute date is shown either in a tooltip or as small text below (like it is currently in the actionlogs / privacy components). This can be deactivated via method parameter.
- If we use the same format for an absolute display as for the full date (DATE_FORMAT_LC6), it is not displayed twice.

I also added unit tests for the new method.

### Testing Instructions

1. Log in to backend.
2. In the user menu (top right), click "Edit Account". In the Tab "Basic Settings", you should find a new parameter "Show Relative Dates in Lists" that defaults to true.
3. Check the following views with this parameter once enabled and once disabled:
   - Module "Latest Actions" on the administrator home dashboard.
   - Module "Logged-in users" on the administrator home dashboard (**attention:** not yet implemented, only after #32911 is merged and I updated this PR here accordingly)
   - Users - User Actions Log
   - Users - Privacy - Requests: You have to create a new request here if there are none yet.
   - Users - Privacy - Consents: You have to give consent with at least one user beforehand. To do this, enable the plugin "System - Privacy Consent" and log in to frontend with a user account.

### Actual result BEFORE applying this Pull Request

![joomla-pr-date-relative-below](https://user-images.githubusercontent.com/1410135/113601787-12a8ea80-9642-11eb-8ad1-55997fa8114c.png)

The dates in the mentioned components and modules are relative for dates which are more recent than a month. For user actions log and the privacy views, the full date is displayed below the relative date.

![joomla-pr-date-relative](https://user-images.githubusercontent.com/1410135/113601785-12105400-9642-11eb-933e-5d46e2d239b6.png)

For the dashboard modules, the full date is not shown anywhere.

### Expected result AFTER applying this Pull Request

#### With "Show Relative Dates in Lists" enabled

![joomla-pr-date-relative-tooltip](https://user-images.githubusercontent.com/1410135/113601781-10df2700-9642-11eb-8010-3a46e4d8585e.png)

In the dashboard modules, the relative dates have a tooltip now showing the full exact date of the action.

In the actions log and privacy views, the display is the same as before.

#### With "Show Relative Dates in Lists" disabled

![joomla-pr-date-absolute-below](https://user-images.githubusercontent.com/1410135/113601807-1a688f00-9642-11eb-9320-7c42a8f6034c.png)

Dates in the mentioned locations are displayed in an absolute format. The full date and time is shown below or in the tooltip, respectively.

### Documentation Changes Required

Maybe we need to document this user parameter somewhere?

### Open Discussion Points

1. Should we include a default setting for users who have not set this parameter by themselves? Or just default to relative display (the way this PR does it now).
2. If date display is set to relative, the date format is DATE_FORMAT_LC6, and the date is more than one month in the past, we get the same date twice. The full date tooltip or div does not provide any additional information in this case. Should we try to detect this case or just leave it that way?
3. Which date format to use for the actions log / privacy components? Currently, it defaults to DATE_FORMAT_LC1, which is "Monday, 05 April 2021". In these list formats, something like DATE_FORMAT_LC5 (2021-05-04 19:00) or DATE_FORMAT_LC6 (2021-05-04 19:00:00) might be better?
4. Is there a chance to get this in 4.0 or should I directly re-target it to 4.1? In my view, #32910 and #32911 (one already merged into 4.0, one RTC) take away some information from these modules. This PR would "fix" this. But I also see that this might count as a new feature, so 4.1 would be okay as well I guess.
